java-checks/src/main/java/org/sonar/java/checks/LeastSpecificTypeCheck.java

```diff
@@ -0,0 +1,224 @@
+      .map(VariableTree::symbol)
+      .filter(p -> p.type().isClass() && !p.type().symbol().isEnum() && !p.type().is("java.lang.String"))
+      .forEach(this::handleParameter);
+  }
+
+  private static boolean isOverloaded(Symbol.MethodSymbol methodSymbol) {
+    return ((Symbol.TypeSymbol) methodSymbol.owner()).lookupSymbols(methodSymbol.name()).size() > 1;
+  }
+
+  private void handleParameter(Symbol parameter) {
+    Type leastSpecificType = findLeastSpecificType(parameter);
+    if (!parameter.type().is(leastSpecificType.fullyQualifiedName())
+      && !leastSpecificType.is("java.lang.Object")) {
+      String suggestedType = leastSpecificType.fullyQualifiedName().replace('$', '.');
+      reportIssue(parameter.declaration(), String.format("Use '%s' here; it is a more general type than '%s'.", suggestedType, parameter.type().name()));
+    }
+  }
+
+  private static Type findLeastSpecificType(Symbol parameter) {
+    List<IdentifierTree> usages = parameter.usages();
+    InheritanceGraph inheritanceGraph = new InheritanceGraph(parameter.type());
+    for (IdentifierTree usage : usages) {
+      Tree parent = usage.parent();
+      while (parent != null && !parent.is(Tree.Kind.ARGUMENTS, Tree.Kind.METHOD_INVOCATION, Tree.Kind.EXPRESSION_STATEMENT, Tree.Kind.FOR_EACH_STATEMENT)) {
+        parent = parent.parent();
+      }
+      if (parent == null || parent.is(Tree.Kind.EXPRESSION_STATEMENT, Tree.Kind.ARGUMENTS)) {
+        return parameter.type();
+      }
+      if (parent.is(Tree.Kind.FOR_EACH_STATEMENT)) {
+        // find Iterable.iterator() method
+        Collection<Symbol> symbols = parameter.type().symbol().lookupSymbols("iterator");
+        symbols.stream()
+          .filter(Symbol::isMethodSymbol)
+          .filter(s -> ((Symbol.MethodSymbol) s).parameterTypes().isEmpty())
+          .findFirst()
+          .ifPresent(inheritanceGraph::update);
+      } else if (parent.is(Tree.Kind.METHOD_INVOCATION)) {
+        MethodInvocationTree mit = (MethodInvocationTree) parent;
+        if (isMethodInvocationOnParameter(parameter, mit)) {
+          inheritanceGraph.update(mit.symbol());
+        }
+      }
+    }
+    return inheritanceGraph.leastSpecificType();
+  }
+
+  private static class InheritanceGraph {
+
+    private final Type startType;
+    private List<List<Type>> chains = null;
+
+    private InheritanceGraph(Type type) {
+      startType = type;
+    }
+
+    private void update(Symbol m) {
+      if (chains == null) {
+        chains = computeChains(m, startType);
+      } else {
+        refineChains(m);
+      }
+    }
+
+    private List<List<Type>> computeChains(Symbol m, Type type) {
+      Symbol.TypeSymbol typeSymbol = type.symbol();
+      Set<ClassJavaType> superTypes = ((JavaSymbol.TypeJavaSymbol) typeSymbol).directSuperTypes();
+      List<List<Type>> result = new ArrayList<>();
+      for (Type superType: superTypes) {
+        for (List<Type> chain: computeChains(m, superType)) {
```

Comment:
This should be dealt with Stream API IMO : supertypes.stream.map(computeChains).map(c -> newChain).collect